src/input/return_enums.rs

```diff
@@ -233,14 +242,190 @@ impl<'a> GenericCollection<'a> {
 #[cfg_attr(debug_assertions, derive(Debug))]
 pub enum GenericMapping<'a> {
     PyDict(&'a PyDict),
+    PyMapping(&'a PyMapping),
     PyGetAttr(&'a PyAny),
     JsonObject(&'a JsonObject),
 }
 
 derive_from!(GenericMapping, PyDict, PyDict);
+derive_from!(GenericMapping, PyMapping, PyMapping);
 derive_from!(GenericMapping, PyGetAttr, PyAny);
 derive_from!(GenericMapping, JsonObject, JsonObject);
 
+pub struct DictGenericIterator<'py> {
+    dict_iter: PyDictIterator<'py>,
+}
+
+impl<'py> DictGenericIterator<'py> {
+    pub fn new(dict: &'py PyDict) -> ValResult<'py, Self> {
+        Ok(Self { dict_iter: dict.iter() })
+    }
+}
+
+impl<'py> Iterator for DictGenericIterator<'py> {
+    type Item = ValResult<'py, (&'py PyAny, &'py PyAny)>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.dict_iter.next().map(Ok)
```

Comment:
It seems that every element has been changed from `(&PyAny, &PyAny)` to `Result<(&PyAny, &PyAny), E>`. I am not sure if it takes time to extract `T` from `Result<T, E>`. Does it break the Zero Cost Abstractions rule?

Reply:
My guess it is that it will not have zero cost, but it will have a tiny cost.

CodSpeed which is now running out benchmarks synthetically with ValGrind should be able to tell us if stuff like this causes a slow down.

Unfortunately codspeed is currently failing on pull requests from forks, but @art049 is working on fixing that now.

Reply:
comparing main to this with pytest-speed shows no significant change

```
Benchmarks being compared                                                                              
┏━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃                       ┃  Before                                     ┃  After                        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│  ID                   │  010                                        │  011                          │
│  Branch               │  main                                       │  tushushu-wip-py-mapping      │
│  Commit SHA           │  1900f4a                                    │  05f676c                      │
│  Commit Message       │  add s390x & ppc64le linux binaries (#333)  │  fix for pypy, small cleanup  │
│                       │                                             │                               │
│                       │  * add s390x & ppc64le linux binaries       │                               │
│                       │                                             │                               │
│                       │  * bump                                     │                               │
│                       │                                             │                               │
│                       │  * setting interpreter and container        │                               │
│  Benchmark Timestamp  │  13:53 (1 mins ago)                         │  13:54 (12 seconds ago)       │
└───────────────────────┴─────────────────────────────────────────────┴───────────────────────────────┘

Benchmarks Comparison                                                                                      
┏━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━┓
┃  Group                 ┃  Benchmark                 ┃  Before (ns/iter)  ┃  After (ns/iter)  ┃  Change  ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━┩
│  create small model    │  small_class_core_dict     │             279.2  │            273.2  │  -2.14%  │
├────────────────────────┼────────────────────────────┼────────────────────┼───────────────────┼──────────┤
│  List[TypedDict]       │  list_of_dict_models_core  │           12470.0  │          12377.2  │  -0.74%  │
├────────────────────────┼────────────────────────────┼────────────────────┼───────────────────┼──────────┤
│  Dict                  │  dict_of_ints_core         │          347572.6  │         343434.7  │  -1.19%  │
├────────────────────────┼────────────────────────────┼────────────────────┼───────────────────┼──────────┤
│  Dict                  │  dict_of_any_core          │          134555.5  │         135946.4  │  +1.03%  │
├────────────────────────┼────────────────────────────┼────────────────────┼───────────────────┼──────────┤
│  Dict JSON             │  dict_of_ints_core_json    │          367568.2  │         368587.2  │  +0.28%  │
├────────────────────────┼────────────────────────────┼────────────────────┼───────────────────┼──────────┤
│  List[DictSimpleMode]  │  many_models_core_dict     │          136714.7  │         137600.5  │  +0.65%  │
└────────────────────────┴────────────────────────────┴────────────────────┴───────────────────┴──────────┘
```

Reply:
Thanks for the explanation, and it seems to be worthy to make a trade off between the speed and code abstraction.